sys/tracefd/tracefd.c

```diff
@@ -0,0 +1,229 @@
+/* System includes */
+#include <stdarg.h>
+#include <stdlib.h>
+#include <string.h>
+
+/* RIOT includes */
+#include "mutex.h"
+#include "xtimer.h"
+
+#ifndef BOARD_NATIVE
+#  include "mtd_sdcard.h"
+#  include "sdcard_spi.h"
+#  include "sdcard_spi_params.h"
+#  include "fs/fatfs.h"
+#endif
+
+#include "tracefd.h"
+
+#define ENABLE_DEBUG (0)
+#include "debug.h"
+
+#ifndef TRACEFD_ROOT_DIR
+#  ifdef BOARD_NATIVE
+#    define TRACEFD_ROOT_DIR "/tmp"         /**< Base directory of trace files (local directory) */
+#  else /* ! BOARD_NATIVE */
+#    define TRACEFD_ROOT_DIR "/sdcard"      /**< Base directory of trace files (SDCARD mount point) */
+#  endif /* BOARD_NATIVE*/
+#endif
+
+#ifndef TRACEFD_MAX_PATH
+#  define TRACEFD_MAX_PATH 256         /**< Maximum filename length */
+#endif
+
+#ifndef TRACEFD_FLUSH_INTERVAL
+#  define TRACEFD_FLUSH_INTERVAL 1000  /**< Interval in ms between two flush of all files */
+#endif
+
+/**
+ * @brief   Trace file descriptor parameters
+ */
+typedef struct {
+    FILE *fd;                           /**< file descriptor */
+    char filename[TRACEFD_MAX_PATH];    /**< filename */
+    mutex_t lock;                       /**< lock protecting file access */
+} tracefd_context_t;
+
+/* Allocate memory for the tracefd contexts */
+tracefd_context_t tracefd_contexts[TRACEFD_NUMOF];
+
+/* Number of initialized tracefd */
+tracefd_t tracefd_initialized = 0;
+
+/* Files flusher thread PID */
+kernel_pid_t files_flusher_pid = -1;
+
+/* Files flusher thread priority */
+#define FILES_FLUSHER_PRIO (THREAD_PRIORITY_MAIN - 1)
+
+/* Files flusher thread stack */
+static char files_flusher_thread_stack[THREAD_STACKSIZE_DEFAULT];
+
+/* Flag set to true to make files flusher thread sleep */
+bool sleep_files_flusher_thread = false;
+
+#ifndef BOARD_NATIVE
+/* This is provided by the sdcard_spi driver
+ * See RIOT/drivers/sdcard_spi/sdcard_spi.c */
+extern sdcard_spi_t sdcard_spi_devs[ARRAY_SIZE(sdcard_spi_params)];
+
+/* Configure MTD device for the first SD card */
+static mtd_sdcard_t mtd_sdcard_dev = {
+    .base = {
+        .driver = &mtd_sdcard_driver
+    },
+    .sd_card = &sdcard_spi_devs[0],
+    .params = &sdcard_spi_params[0],
+};
+
+/* File system specific descriptor */
+static fatfs_desc_t fs_desc = {
+    .vol_idx = 0,     /* Low level device that is used by FatFs */
+    .fat_fs = {
+        .fs_type = FS_EXFAT
+    }
+};
+
+/* Declare mtd devices (for use within diskio layer of fatfs) */
+/* Declared as `extern` in `RIOT/pkg/fatfs/fatfs_diskio/mtd/mtd_diskio.c` */
+mtd_dev_t *fatfs_mtd_devs[FF_VOLUMES];
+
+/* Define the vfs mount point */
+static vfs_mount_t flash_mount = {
+    .fs = &fatfs_file_system,
+    .mount_point = TRACEFD_ROOT_DIR,
+    .private_data = &fs_desc,
+};
+#endif /* BOARD_NATIVE */
```

Comment:
Why is it needed ?

Reply:
I removed #ifdef and split this into a common file and files specific to the target.